src/System.Linq/src/System/Linq/Enumerable.SelectWithCountIterator.cs

```diff
@@ -0,0 +1,62 @@
+﻿
+using System.Collections.Generic;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        internal abstract class SelectWithCountIterator<TSource> : SelectIterator<TSource>, ICollection<TSource>
+        {
+            public override abstract Iterator<TSource> Clone();
+
+            public override abstract bool MoveNext();
+
+            public override abstract IEnumerable<TSource> Where(Func<TSource, bool> predicate);
+
+            public override abstract IEnumerable<TResult> SelectImpl<TResult>(Func<TSource, TResult> selector);
+
+            #region ICollection<TSource> implementation
+
+            public abstract int Count { get; }
+
+            public bool IsReadOnly { get { return true; } }
+
+            public void Add(TSource item)
+            {
+                throw NotImplemented.ByDesign;
+            }
+
+            public void Clear()
+            {
+                throw NotImplemented.ByDesign;
+            }
+
+            public bool Contains(TSource item)
+            {
+                throw NotImplemented.ByDesign;
```

Comment:
The more I think about it the more I tend to lean toward a big change to the entire perf improvement concept. Instead of relying on publicly available `ICollection<T>` which contains much more members than necessary maybe we should create new, internal, non variant interface that would be shared between `System.Collections.Generic` and `System.Linq` and only contain `Count` property - that would be the interface Iterators like `SelectCollectionIterator` would implement and `List<T>` would use to make list initialization optimizations.

This would solve multiple issues:
- iterators returned by LINQ method wouldn't be castable to new interfaces, because the interface would be internal
- you wouldn't have to implement `Contains` method which basically duplicates code from `Enumerable.Contains` without `ICollection<T>` check
- cast check would be much faster than to `IReadOnlyCollection<T>`, which is slow because it's covariant

The drawback is, you'd have to wait for `System.Collections.Generic` to be open sourced to proceed with the change, because of changes needed in `List<T>` and other collections.